pilot/pkg/model/push_context.go

```diff
@@ -78,9 +80,127 @@ type PushContext struct {
+	// The IP is used because K8S Endpoints object associated with a Service only include the IP.
+	// We use Endpoints to track the membership to a service and readiness.
+	WorkloadUpdate(id string, labels map[string]string, annotations map[string]string)
+}
+
+// IstioEndpoint has the information about a single address+port for a specific
+// service and shard. Unlike ServiceInstance, it has less info and require less
+// allocation.
+type IstioEndpoint struct {
+
+	// Labels points to the workload or deployment labels.
+	Labels map[string]string
+
+	// Family indicates what type of endpoint, such as TCP or Unix Domain Socket.
+	// Default is TCP.
+	Family AddressFamily
+
+	// Address is the address of the endpoint, using envoy proto.
+	Address string
+
+	// EndpointPort is the port where the workload is listening, can be different
+	// from the service port.
+	EndpointPort uint32
+
+	// ServicePortName tracks the name of the port, to avoid 'eventual consistency' issues.
+	// Sometimes the Endpoint is visible before Service - so looking up the port number would
+	// fail. Instead the mapping to number is made when the clusters are computed. The lazy
+	// computation will also help with 'on-demand' and 'split horizon' - where it will be skipped
+	// for not used clusters or endpoints behind a gate.
+	ServicePortName string
+
+	// UID identifies the workload, for telemetry purpose.
+	UID string
+
+	// EnvoyEndpoint is a cached LbEndpoint, converted from the data, to
+	// avoid recomputation
+	EnvoyEndpoint *endpoint.LbEndpoint
+
+	// ServiceAccount holds the associated service account.
+	// May be replaced once a better optimization/struct is available for secure naming.
+	ServiceAccount string
+}
+
+// ServiceShards holds the set of endpoint shards of a service. Registries update
+// individual shards incrementally. The shards are aggregated and split into
+// clusters when a push for the specific cluster is needed.
+type ServiceShards struct {
+
+	// Shards is used to track the shards. EDS updates are grouped by shard.
+	// Current implementation uses the registry name as key - in multicluster this is the
+	// name of the k8s cluster, derived from the config (secret).
+	Shards map[string]*EndpointShard
+
+	// ServiceAccounts has the concatenation of all service accounts seen so far in endpoints.
+	// This is updated on push, based on shards. If the previous list is different than
+	// current list, a full push will be forced, to trigger a secure naming update.
+	// Due to the larger time, it is still possible that connection errors will occur while
+	// CDS is updated.
+	ServiceAccounts map[string]bool
+}
+
+// EndpointShard contains all the endpoints for a single shard (subset) of a service.
+// Shards are updated atomically by registries. A registry may split a service into
+// multiple shards (for example each deployment, or smaller sub-sets).
+type EndpointShard struct {
```

Comment:
What is a VNID ? This has nothing to do with networks - it's just a grouping of endpoints that can be further optimized.

Reply:
Yes, but is it an arbitrary grouping? The code is structured such that each registry (each k8s cluster, or consul/ vm/cloud foundry) gets to update its own list of endpoints such that you can propagate just the requisite changes.. The VNID is essentially that. a network boundary established by the orchestration platform. There is no need to create yet another abstraction called network that creates yet another possible grouping of endpoints. VNID is virtual network ID. Essentially, its just the cluster name.. Atleast, name this as a cluster name. SO that it aligns well with the existing multicluster stuff that is already keying things by the cluster name.

Reply:
This grouping has nothing to do with 'VNID' or networks. And I agree, no need to create a new abstraction to group endpoints, just an index to efficiently compute - in the PR that deals with that.

For example: consul may aggregate data from multiple k8s and non-k8s 'clusters', and VMs - and all of them can be spread in multiple networks. The consul implementation will group the updates, to avoid very large lists and allocations. Consul doesn't define 'network boundaries'. Neither does k8s, in a large multi-region k8s the endpoints may be in different regions - and user may want to group them by locality. 

We will introduce VNID - and a separate data structure optimized for split horizon - but again that's orthogonal to this PR.

Reply:
and how will this network level grouping differ from the exisitng consul level grouping? 
And this example of consul aggregating from multiple places - implies that the services that exist there, are functioning well *without Istio*, i.e. can talk to one another already. Ergo, after Installing Istio, these services can continue talking to each other in the same way they talked before Istio, without any sophisticated gateway for split-horizon. (All we need here is zone-aware routing, which is not to be conflated with split-horizon).  And istio will continue to treat the source from Consul as one registry (doesn't care if its aggregating from k8s/cf). 

You use the split-horizon only when endpoints from one registry cannot reach the endpoints in another registry. And this is the situation we currently have and the endpoint grouping you did is essentially solving that. So, by adding any other index/abstraction or whatever for VNID, you end up creating more confusion because its not really clear how endpoints can't communicate before the mesh but can with the mesh on the same network.